include/swift/Sema/ConstraintSystem.h

```diff
@@ -3766,7 +3766,10 @@ class ConstraintSystem {
                                        ConstraintLocatorBuilder locator,
                                        ConstraintFix *compatFix = nullptr);
   
-  /// Add a materialize constraint for a pack expansion.
+  /// For a given `patternType`, materialize the pack expansion and return a type variable with the result.
+  ///
+  /// \param patternType A single unlabeled tuple with pack expansion.
+  /// \param locator The locator.
```

Comment:
Previous discussion: https://github.com/apple/swift/pull/67594#discussion_r1418549390

> This explains what happens inside but I'm not sure whether the users should care. I think all the care about that given a `patternType` that is a single unlabeled tuple with pack expansion, materialize pack expansion and store the result into returned type variable...

Is such a distorted and possibly misleading description worth it? `addConstraint` methods and type variable types are public. What is the merit in concealing that it adds a constraint?
___

* Could we rename `patternType` to `tupleType`?
* Could we add a `assert(isSingleUnlabeledPackExpansionTuple(tupleType)` assertion to the implementation?
* ```suggestion
    /// For a given `patternType`, materialize the pack expansion and return a type variable with the result.
    ///
    /// \param tupleType A tuple containing a single unlabeled element that is either a pack expansion or a type variable constrained to a pack expansion.
    /// \param locator A locator associated with the tuple.
  ```

Reply:
cc @xedin

Reply:
All I'm trying to point out is that the constraint itself an implementation detail, the comment should explain what inputs the method expects to give users clear understanding of invariants, what operations are performed, and what result represents. If we want to mention constraint it's okay to reference it in addition with a `\c` somehow.

Probably something like the following should be reasonable enough: Given a tuple with a single unlabeled element that represents a pack expansion (either directly via \c PackExpansionType or through a type variable constrained to a pack expansion), materialize the pack expansion and return its pattern type as a result. The result is a type variable because element of the tuple is not required to be resolved at the time of the call and operation is delayed until the element is sufficiently resolved (see \c simplifyMaterializePackExpansionConstraint).

Reply:
I think my problem is that "materialize the pack expansion and return its pattern type" sounds ambiguous without a mention of the constraint in my mind. If you reckon the suggested comment captures the difference between simply returning the tuple element type and adding a constraint, I’m quite happy with it. Sorry if my choice of words made it feel like I was looking for an argument. FWIW.

Reply:
Constraint is not always required, only when the pattern is not yet resolved but I am not sure whether it makes any difference to the users of the API.